core/src/main/scala/org/apache/spark/resource/ResourceAllocator.scala

```diff
@@ -20,6 +20,49 @@ package org.apache.spark.resource
 import scala.collection.mutable
 
 import org.apache.spark.SparkException
+import org.apache.spark.resource.ResourceAmountUtils.ONE_ENTIRE_RESOURCE
+
+private[spark] object ResourceAmountUtils {
+  /**
+   * Using "double" to do the resource calculation may encounter a problem of precision loss. Eg
+   *
+   * scala&gt; val taskAmount = 1.0 / 9
+   * taskAmount: Double = 0.1111111111111111
+   *
+   * scala&gt; var total = 1.0
+   * total: Double = 1.0
+   *
+   * scala&gt; for (i &lt;- 1 to 9 ) {
+   * |   if (total &gt;= taskAmount) {
+   * |           total -= taskAmount
+   * |           println(s"assign $taskAmount for task $i, total left: $total")
+   * |   } else {
+   * |           println(s"ERROR Can't assign $taskAmount for task $i, total left: $total")
+   * |   }
+   * | }
+   * assign 0.1111111111111111 for task 1, total left: 0.8888888888888888
+   * assign 0.1111111111111111 for task 2, total left: 0.7777777777777777
+   * assign 0.1111111111111111 for task 3, total left: 0.6666666666666665
+   * assign 0.1111111111111111 for task 4, total left: 0.5555555555555554
+   * assign 0.1111111111111111 for task 5, total left: 0.44444444444444425
+   * assign 0.1111111111111111 for task 6, total left: 0.33333333333333315
+   * assign 0.1111111111111111 for task 7, total left: 0.22222222222222204
+   * assign 0.1111111111111111 for task 8, total left: 0.11111111111111094
+   * ERROR Can't assign 0.1111111111111111 for task 9, total left: 0.11111111111111094
+   *
+   * So we multiply ONE_ENTIRE_RESOURCE to convert the double to long to avoid this limitation.
+   * Double can display up to 16 decimal places, so we set the factor to
+   * 10, 000, 000, 000, 000, 000L.
+   */
+  final val ONE_ENTIRE_RESOURCE: Long = 10000000000000000L
```

Comment:
This is way too hacky.
I get it, but, the right solution in your example is to ask for something like 0.111 GPUs to get 9. This works fine.

Reply:
I agree it also doesn't feel wholly satisfying. In this example all of those values work as 9 times even 0.11 leaves you with less than 0.11 remaining, so you schedule 9. It would also imply there is 0.01 GPU left when that isn't the intent. In practice, I strongly doubt anyone is ever scheduling, let's say, more than 100 tasks on one GPU.

(But what about non-GPU resources? there aren't any now. Are there resources you'd schedule very very small fractions of? I can't think of any even in the future.)

Going down the rabbit hole of floating-point precision, I think we hit that no matter what. If I ask for 1e-16 resources, any way we interpret that probably is slightly imprecise as it's interpreted as float somewhere. But these are unrealistic use cases.

Reply:
@srowen  what if your major concern here?  Are you seeing performance issues?  just more complex then you think is necessary?   i don't think its hacky, it might be overkill but I do agree its likely more then we need for precision.  I would go more then 100 though as 100 tasks per some resource seems unlikely but not impossible, 1000 - 10000 much more unlikely.

Reply:
I don't think we can solve floating-point accuracy here in the general case, and this will virtually never arise anyway, except in one important class of case -- n GPUs where n < 10 and n is relatively prime to 10. Like, 3 even. A person writing down the resource utilization will almost surely write "0.333", but one can imagine supplying `str(1./3.)` programmatically. And then this issue could arise. Some string like "0.333333333" may end up as a float that has a value just over 1/3.

The other issue is approaching it this way by just multiplying by a long. I guess instead this can be done properly with BigDecimal at least? and I don't think this should touch so many parts of the code. Surely this just affects how the resource request string is parsed and compared to available resources

Reply:
>  I don't think this should touch so many parts of the code. Surely this just affects how the resource request string is parsed and compared to available resources

I believe the majority of the changes are still required here.  This PR is making it possible so that different tasks could use different resources per task. The previous way we tracked addresses won't work because it statically created an array of addressed based on a static configuration (ie all tasks get the same resources per task).  To properly support different tasks asking for different resource the changes in this pr now pass around the value of what is left.   If I'm misunderstanding what you are saying please clarify.

Reply:
OK that part makes sense if it's kind of separate from the rounding issue

Reply:
Hi @srowen, Yeah, as long as we use double to do the calculation, we can't avoid the precision issue.


``` scala
scala> val taskAmount = 0.05
taskAmount: Double = 0.05

scala> var total = 1.0
total: Double = 1.0

scala> for (i <- 1 to 20 ) {
     |   if (total >= taskAmount) {
     |              total -= taskAmount
     |              println(s"assign $taskAmount for task $i, total left: $total")
     |      } else {
     |              println(s"ERROR Can't assign $taskAmount for task $i, total left: $total")
     |      }
     | }
assign 0.05 for task 1, total left: 0.95
assign 0.05 for task 2, total left: 0.8999999999999999
assign 0.05 for task 3, total left: 0.8499999999999999
assign 0.05 for task 4, total left: 0.7999999999999998
assign 0.05 for task 5, total left: 0.7499999999999998
assign 0.05 for task 6, total left: 0.6999999999999997
assign 0.05 for task 7, total left: 0.6499999999999997
assign 0.05 for task 8, total left: 0.5999999999999996
assign 0.05 for task 9, total left: 0.5499999999999996
assign 0.05 for task 10, total left: 0.4999999999999996
assign 0.05 for task 11, total left: 0.4499999999999996
assign 0.05 for task 12, total left: 0.39999999999999963
assign 0.05 for task 13, total left: 0.34999999999999964
assign 0.05 for task 14, total left: 0.29999999999999966
assign 0.05 for task 15, total left: 0.24999999999999967
assign 0.05 for task 16, total left: 0.19999999999999968
assign 0.05 for task 17, total left: 0.1499999999999997
assign 0.05 for task 18, total left: 0.09999999999999969
assign 0.05 for task 19, total left: 0.049999999999999684
ERROR Can't assign 0.05 for task 20, total left: 0.049999999999999684
```

Take the above code as an example, taskAmount=0.05, which I think is a very normal value, but we can't assign the resources to a total of 20 tasks due to precision issues. 

If we convert double to Long, it can work

``` scala
scala> val ONE_ENTIRE_RESOURCE: Long = 10000000000000000L
ONE_ENTIRE_RESOURCE: Long = 10000000000000000

scala> 

scala> val taskAmount = 0.05
taskAmount: Double = 0.05

scala> var total = 1.0 * ONE_ENTIRE_RESOURCE
total: Double = 1.0E16

scala> for (i <- 1 to 20 ) {
     |   if (total >= taskAmount * ONE_ENTIRE_RESOURCE) {
     |              total -= taskAmount * ONE_ENTIRE_RESOURCE
     |              println(s"assign $taskAmount for task $i, total left: $total")
     |      } else {
     |              println(s"ERROR Can't assign $taskAmount for task $i, total left: $total")
     |      }
     | }
assign 0.05 for task 1, total left: 9.5E15
assign 0.05 for task 2, total left: 9.0E15
assign 0.05 for task 3, total left: 8.5E15
assign 0.05 for task 4, total left: 8.0E15
assign 0.05 for task 5, total left: 7.5E15
assign 0.05 for task 6, total left: 7.0E15
assign 0.05 for task 7, total left: 6.5E15
assign 0.05 for task 8, total left: 6.0E15
assign 0.05 for task 9, total left: 5.5E15
assign 0.05 for task 10, total left: 5.0E15
assign 0.05 for task 11, total left: 4.5E15
assign 0.05 for task 12, total left: 4.0E15
assign 0.05 for task 13, total left: 3.5E15
assign 0.05 for task 14, total left: 3.0E15
assign 0.05 for task 15, total left: 2.5E15
assign 0.05 for task 16, total left: 2.0E15
assign 0.05 for task 17, total left: 1.5E15
assign 0.05 for task 18, total left: 1.0E15
assign 0.05 for task 19, total left: 5.0E14
assign 0.05 for task 20, total left: 0.0
```

Reply:
TL;DR I understand the issue with floating-point numbers. Using a long doesn't totally fix every theoretical case but BigDecimal should. Can you make this work with BigDecimal?

Reply:
Hi @srowen, Thx for your suggestion, Yeah, definitely, I can try BigDecimal, But I would also like to learn which case the "using the Long" can't resolve.  Thx.

Reply:
This doesn't work, for example, if your taskAmount is 1.0/11.0

Reply:
> Sometimes 1/n as a float is bigger than 1/n, sometimes smaller, and the problem occurs when it's bigger

Yeah, Good point. I also thought about this issue when doing this PR. The principle of this PR is to use the *final value* of 1/n to do the calculation, if it's bigger than the real "1/n", that's fine, spark will only assign the resources to *"n-1"* tasks instead of *n* tasks. Because Spark thinks the "final bigger 1/n" is the final value the user would like to set. What Spark should do is do the correct calculation based on the final "1/n" no matter if it's bigger or smaller.

Reply:
You can decide that the excess request is small enough (<1/n) and allow it, but it feels complex compared to representing the user's input string exactly as BigDecimal and just making the natural comparison to remaining resources

Reply:
Hi @srowen,

The **using Long** way is quite the same with the BigDecimal. The default scale of BigDecimal is 16, so this PR chooses (ONE_ENTIRE_RESOURCE  = 1E16.toLong)

``` scala
scala>     val ONE_ENTIRE_RESOURCE: Long = 1E16.toLong
     |     val taskAmount = 0.3333333333333334
     | 
     |     val usingLong = (taskAmount * ONE_ENTIRE_RESOURCE).toLong
     | 
     |     val bigDec = BigDecimal(taskAmount).toDouble
val ONE_ENTIRE_RESOURCE: Long = 10000000000000000
val taskAmount: Double = 0.3333333333333334
val usingLong: Long = 3333333333333334
val bigDec: Double = 0.3333333333333334
``` 

So if we need to ensure the input is small enough (<1/n) and we can set the scale to be like 14 for BigDecimal, and similarly, to keep align with BigDecimal, we can set `ONE_ENTIRE_RESOURCE  = 1E14.toLong`

``` scala
scala>     import scala.math.BigDecimal.RoundingMode
     | 
     |     val ONE_ENTIRE_RESOURCE: Long = 1E14.toLong
     |     val taskAmount = 0.3333333333333334
     | 
     |     val usingLong = (taskAmount * ONE_ENTIRE_RESOURCE).toLong
     | 
     |     val bigDec = BigDecimal(taskAmount).setScale(14, RoundingMode.DOWN).toDouble
import scala.math.BigDecimal.RoundingMode
val ONE_ENTIRE_RESOURCE: Long = 100000000000000
val taskAmount: Double = 0.3333333333333334
val usingLong: Long = 33333333333333
val bigDec: Double = 0.33333333333333
```

Reply:
No. The point of Big decimal is that you never have to use double or float. This isn't actually doing math with BD

Reply:
Hi @srowen. Actually, this PR converts the taskAmount which is a double/float value to a Long by multiplying `1E16.toLong`, and then the following calculation is based on Long instead of double/float. you can see, all the APIs of [ExecutorResourcesAmount](https://github.com/apache/spark/pull/44690/files#diff-91f559d0e4aeff4052cd7c6575d8622d911ecad0852040f64ccc1d3abfd605faR41) are using Long. Even the assigned resource of a task is still keeping the Long, you can refer to [there](https://github.com/apache/spark/pull/44690/files#diff-381b69941735d860cddcbb3b565abfef591a890aadecbc235283f3a57de9ba61R62).

Yeah, but you think we should use BigDecimal, I'm Okay for that, I can make a PR for master branch first, and then cherry-pick to 3.5 branch.

Reply:
I know. Above I give an example where multiplying by long works. I'm referring to your example of BigDecimal above, which does not use (only) BigDecimal. Please just try it.

Reply:
Hi @srowen. Sure, let me have a PR using BD for master branch. Thx